internal/restic/repository.go

```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 
 	"github.com/restic/restic/internal/crypto"
+	"github.com/restic/restic/internal/ui/progress"
```

Comment:
I think this is a sign that (after this PR) we should drop this file and the [single-implementation interfaces](https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a) in it, and just let client code work on *repository.Repository instead. Changes to that type already require changes here, and now every package that wants to handle a restic.ID drags in ui/progress. That's not a very big deal yet, but it does mean that ui/progress cannot be told about restic.ID when the need arises.

Reply:
We currently have a few tests that wrap `restic.Repository` for the checker and the archiver. These wouldn't be possible anymore if we drop the interface. Although the wrapper repos only use `LoadTree`, `LoadBlob` and `LookupBlobSize`. Replacing these wrappers with different interfaces would probably be rather messy.

Dragging in the progress.Counter is a bit unfortunate, but if necessary we can decouple that with an additional interface ;-) .

I also plan to split the restic package into a backend part, the snapshot data model part and the repository related parts. There's a bunch of totally experimental commits at https://github.com/MichaelEischer/restic/tree/layered-repo2 which includes the just mentioned refactoring and a rather random collection of other stuff.

Reply:
If we define checker.Repository and archiver.Repository with the set of operations those packages need, then IMO that's not messy at all. It signals that checker needs one set of operations and archiver another. There's a motto "[accept interfaces, return structs](https://nathanleclaire.com/blog/2015/10/10/interfaces-and-composition-for-effective-unit-testing-in-golang/)", and although it's [controversial](https://github.com/go-proverbs/go-proverbs.github.io/issues/37), I think it applies here.

This is all beside the point, which is that your PR LGTM. Testing the waters for a further refactor, which I'll postpone. Looking forward to your follow-up :)

Reply:
Cleaning up all my old beginner mistakes, awesome! :)